src/radical/pilot/agent/agent_0.py

```diff
@@ -446,27 +446,33 @@ def _state_cb_of_services(self, topic, msg):
                 uid   = task['uid']
                 state = task['state']
 
-                if uid in self._task_service_data:
-                    # update task info and signal task service thread
-                    self._log.debug('unlock 2 %s', uid)
-                    self._task_service_data[uid][1] = task
-                    self._task_service_data[uid][0].set()
-
-
-        elif cmd == 'update':
-
-            for thing in ru.as_list(arg):
-
-                uid   = thing['uid']
-                state = thing['state']
-
-                if uid in self._workers:
-                    if state == rps.AGENT_STAGING_OUTPUT:
-                        with self._lock:
-                            self._workers[uid]['state'] = 'DONE'
-
-            self.state_cb(ru.as_list(arg))
+                if state == rps.AGENT_EXECUTING:
+                    print("will remove this")
+                    self._log("Service x is started. please pass some ID.")
+                else:
+                    pass
+                    # TODO We should do something.
+
+                # if uid in self._task_service_data:
+                #     # update task info and signal task service thread
+                #     self._log.debug('unlock 2 %s', uid)
+                #     self._task_service_data[uid][1] = task
+                #     self._task_service_data[uid][0].set()
+        # Waiting for every callback from every service, and then notify the main thread.
+        e.wait()
```

Comment:
I have used the Event class but no proper multithreading in python. 
Want to wanted to do - 


```
_start_services()
threading
... spawn threads
... event 
... event.set() //blcoked until all threads are done


thread_method()
... start and check service for a particular
... e.wait()
```

Reply:
Ah, now I see what you are trying.  Allow me to suggest a simpler path.

In line 364, you advance the service tasks and push them to the stager / scheduler / executor components.  Right after that `advance` call is when you want to wait until the tasks are executing, so you can add there something like 

```py
self._service_wait_event.wait()
```

(possibly with some timeout and error check).

Here, in `_state_cb_of_services`, you will get state notifications for all tasks.  There is no need to issue your own events (`service_state_update`) - the executor component will send a state update already when the service tasks are entering `AGENT_EXECUTING` state.  You pasted the code from `https://github.com/radical-cybertools/radical.pilot/blob/devel/src/radical/pilot/raptor/master.py#L245` which I referred to earlier in the comment below - that is the code you should actually use (just the state needs changing).

So you basically just have to count the state update messages - if you start 3 service tasks, you should get 3 state update messages there (to be safe, you can also check the task IDs to make sure you check the right messages).  Once you received all three it would be save to set
```py
self._service_wait_event.set()
```
to let the other thread (which does the `...wait()` above) continue.

```py
if cmd == 'update':
    for thing in ru.as_list(arg):
        uid   = thing['uid']
        state = thing['state']
        if uid in self._service_task_ids and state == rps.AGENT_EXECUTING:
            self._active_services += 1
            if self._active_services == len(self._service_task_ids):
            self._service_wait_event.set()
```
or something like that...

Reply:
made changes. please see